plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSig/Stage0.hs

```diff
@@ -0,0 +1,247 @@
+                let numSigned = $$(P.length) ($$(P.filter) (\pk -> $$containsPk pk pks) signatories)
+                in $$(P.geq) numSigned numReq
+        in proposalAccepted'
+    ||]
+
+-- | @valuePreserved v p@ is true if the pending transaction @p@ pays the amount
+--   @v@ to a single pay-to-script output at this script's address.
+valuePreserved :: Q (TExp (Value -> PendingTx -> Bool))
+valuePreserved = [||
+
+        let valuePreserved' :: Value -> PendingTx -> Bool
+            valuePreserved' vl ptx = 
+                let ownHash = $$(Validation.ownHash) ptx
+                    numOutputs = $$(P.length) ($$(Validation.scriptOutputsAt) ownHash ptx)
+                    valueLocked = $$(Validation.valueLockedBy) ptx ownHash
+                in $$(P.and) ($$(P.eq) 1 numOutputs) ($$(Value.eq) valueLocked vl)
+        in valuePreserved'
+
+    ||]
+
+-- | @valuePaid pm ptx@ is true if the pending transaction @ptx@ pays
+--   the amount specified in @pm@ to the public key address specified in @pm@
+valuePaid :: Q (TExp (Payment -> PendingTx -> Bool))
+valuePaid = [||
+
+        let valuePaid' :: Payment -> PendingTx -> Bool
+            valuePaid' (Payment vl pk _) ptx = $$(Value.eq) vl ($$(Validation.valuePaidTo) ptx pk)
+        in valuePaid'
+    ||]
+
+-- | Equality of 'Payment' values
+paymentEq :: Q (TExp (Payment -> Payment -> Bool))
+paymentEq = [||
+
+        let paymentEq' :: Payment -> Payment -> Bool
+            paymentEq' (Payment vl pk sl) (Payment vl' pk' sl') = 
+                $$(P.and) ($$(P.and) ($$(Value.eq) vl vl') ($$(Validation.eqPubKey) pk pk')) ($$(Slot.eq) sl sl')
+        in paymentEq'
+
+    ||]
+
+-- | Equality of 'State' values
+stateEq :: Q (TExp (State -> State -> Bool))
+stateEq = [||
+        let stateEq' :: State -> State -> Bool
+            stateEq' (InitialState v) (InitialState v') =
+                $$(Value.eq) v v'
+            stateEq' (CollectingSignatures vl pmt pks) (CollectingSignatures vl' pmt' pks') =
+                $$(P.and) ($$(P.and) ($$(Value.eq) vl vl') ($$paymentEq pmt pmt')) ($$pkListEq pks pks')
+            stateEq' _ _ = False
+
+        in stateEq'
+    ||]
+
+-- | @step params state input@ computes the next state given current state
+--   @state@ and the input.
+step :: Q (TExp (Params -> State -> Input -> State))
+step = [||
+
+    let step' :: Params -> State -> Input -> State
+        step' p s i = 
+            case (s, i) of
+                (InitialState vl, ProposePayment pmt) -> 
+                    CollectingSignatures vl pmt []
+                (CollectingSignatures vl pmt pks, AddSignature pk) ->
+                    CollectingSignatures vl pmt (pk:pks)
+                (CollectingSignatures vl pmt _, Cancel) ->
+                    InitialState vl
+                (CollectingSignatures vl pmt@(Payment vp _ _) pks, Pay) ->
+                    let vl' = $$(Value.minus) vl vp in
+                    InitialState vl'
```

Comment:
I wonder whether it would make sense to have some intermediate state like `PaymentPerformed`, maybe for some specific business logic.